### PR TITLE
[Feature] Support C3D

### PR DESCRIPTION
**Pull request by dreamerlin**

### CODE REVIEW

**Comment by dreamerlin** [comment]
The linting failed. Maybe you can try the pre-commit, which use some formatter to correct the format.

**Comment by dreamerlin** [comment]
Kindly ping : )

File: configs/recognition/c3d/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+### UCF-101
+
+| config | resolution | gpus | backbone | pretrain | top1 acc | top5 acc | inference_time(video/s) | gpu_mem(M) | ckpt | log | json |
+|:--|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|
+|[c3d_sports1m_16x1x1_30e_ucf101_rgb.py](/configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py)|128x171|8| c3d | sports1m | 80.04(83.37) | 95.19(96.25) | x | 6053 | [ckpt](https://download.openmmlab.com/mmaction/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb/c3d_sports1m_16x1x1_30e_ucf101_rgb_20201016-e0d4bd99.pth)|[log](https://download.openmmlab.com/mmaction/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb/20201016_101722.log)|[json](https://download.openmmlab.com/mmaction/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb/20201016_101722.log.json)
+
+Notes:
+
+1. The author of C3D used SVM to classify videos on UCF-101, we simply used a linear classifier and the **accuracy in brackets** was tested under the TSN framework with 10 segments.
+2. The **gpus** indicates the number of gpu (32G V100) we used to get the checkpoint. It is noteworthy that the configs we provide are used for 8x4 gpus as default.
```
Comment (dreamerlin) [inline configs/recognition/c3d/README.md:36]: "It is noteworthy that the configs we provide are used for 8x4 gpus as default." wrong, for C3D, it is 8 gpus as default, right ?

File: configs/recognition/c3d/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+According to the [Linear Scaling Rule](https://arxiv.org/abs/1706.02677), you may set the learning rate proportional to the batch size if you use different GPUs or videos per GPU,
+e.g., lr=0.01 for 4 GPUs * 2 video/gpu and lr=0.08 for 16 GPUs * 4 video/gpu.
+3. The **inference_time** is got by this [benchmark script](/tools/analysis/benchmark.py), where we use the sampling frames strategy of the test setting and only care about the model inference time,
+not including the IO time and pre-processing time. For each setting, we use 1 gpu and set batch size (videos per gpu) to 1 to calculate the inference time.
+4. The values in columns named after "reference" are the results got by training on the original repo, using the same model settings.
```
Comment (dreamerlin) [inline configs/recognition/c3d/README.md:41]: This line seems useless for C3D README

File: configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
+# model settings
+model = dict(
+    type='Recognizer2D',
+    backbone=dict(
+        type='C3D',
+        pretrained='https://download.openmmlab.com/mmaction/recognition/c3d'
+        '/c3d_sports1m_pretrain_20201016-dcc47ddc.pth',
```
Comment (dreamerlin) [inline configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:7]: you can use `# noqa: E501` like [this](https://github.com/open-mmlab/mmaction2/blob/master/configs/recognition/csn/ircsn_ig65m_pretrained_bnfrozen_r152_32x2x1_58e_kinetics400_rgb.py#L7-L8)

File: configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        ann_file=ann_file_val,
+        data_prefix=data_root_val,
+        pipeline=val_pipeline),
+    test=dict(
+        type=dataset_type,
+        ann_file=ann_file_test,
+        data_prefix=data_root_val,
+        pipeline=test_pipeline))
+# optimizer
+optimizer = dict(type='SGD', lr=0.001, momentum=0.9, weight_decay=0.0005)
```
Comment (dreamerlin) [inline configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:114]: add a comment `# this lr is used for 8 gpus `

File: mmaction/datasets/pipelines/augmentations.py [hunk 5]
```diff
@@ -971,3 +971,69 @@ def __repr__(self):
...
+        n = len(results['imgs'])
+        h, w, c = results['imgs'][0].shape
+        imgs = np.empty((n, h, w, c), dtype=np.float32)
+
+        if self.mean.shape != (n, h, w, c):
+            raise ValueError(
+                'The shape of mean values and imgs mismatch in pixel-wise')
+
+        imgs = np.subtract(results['imgs'], self.mean)
+        results['imgs'] = imgs
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/augmentations.py:1030]: add
```
results['mean'] = mean
```

**Comment by dreamerlin** [comment]
There are conflicts now.

**Comment by dreamerlin** [comment]
And please click [this](https://cla-assistant.io/open-mmlab/mmaction2?pullRequest=226) to sign the agreement.

File: configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
+# model settings
+model = dict(
+    type='Recognizer2D',
```
Comment (dreamerlin) [inline configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:3]: 2D ?

File: configs/recognition/c3d/README.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+For more details, you can refer to **Training setting** part in [getting_started](/docs/getting_started.md#training-setting).
+
+## Test
+
+You can use the following command to test a model.
+```shell
+python tools/test.py ${CONFIG_FILE} ${CHECKPOINT_FILE} [optional arguments]
+```
+
+Example: test C3d model on UCF-101 dataset and dump the result to a json file.
```
Comment (innerlee) [inline configs/recognition/c3d/README.md:67]: C3D

File: configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
+# model settings
+model = dict(
+    type='Recognizer3D',
+    backbone=dict(
+        type='C3D',
+        pretrained=  # noqa: E251
+        'https://download.openmmlab.com/mmaction/recognition/c3d/c3d_sports1m_pretrain_20201016-dcc47ddc.pth',  # noqa: E501
```
Comment (innerlee) [inline configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:7]: since you are ignoring E501, why not put the above two lines in one line?

Comment (congee524) [reply configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:7]: I put the two lines with reference to "configs/recognition/csn/ircsn_ig65m_pretrained_bnfrozen_r152_32x2x1_58e_kinetics400_rgb": line 7-8.
In fact, when I want to put them in one line, the IDE will automatically wrap them (it follows pep8 code style).

Comment (dreamerlin) [reply configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:7]: And the `# noqa: E251` is used to pass CI

Comment (innerlee) [reply configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:7]: which formatter are used? i think a good formatter should respect the `noqa` marks

File: mmaction/datasets/pipelines/augmentations.py [hunk 9]
```diff
@@ -1264,3 +1264,70 @@ def __repr__(self):
...
+            transpose to reverse or permute the mean values. Default: None.
+    """
+
+    def __init__(self, mean=None, mean_file=None, transpose_axes=None):
+        self.mean = mean
+        self.mean_file = mean_file
+        self.transpose_axes = transpose_axes
+
+        if self.mean_file is not None:
+            self.mean = np.load(self.mean_file)
```
Comment (innerlee) [inline mmaction/datasets/pipelines/augmentations.py:1292]: why load the same file repeatedly?

File: mmaction/datasets/pipelines/augmentations.py [hunk 10]
```diff
@@ -1264,3 +1264,70 @@ def __repr__(self):
...
+
+        n = len(results['imgs'])
+        h, w, c = results['imgs'][0].shape
+        imgs = np.empty((n, h, w, c), dtype=np.float32)
+
+        if self.mean.shape != (n, h, w, c):
+            raise ValueError(
+                'The shape of mean values and imgs mismatch in pixel-wise')
+
+        imgs = np.subtract(results['imgs'], self.mean)
```
Comment (innerlee) [inline mmaction/datasets/pipelines/augmentations.py:1322]: why not use `-`?

File: mmaction/datasets/pipelines/augmentations.py [hunk 11]
```diff
@@ -1264,3 +1264,70 @@ def __repr__(self):
...
+
+        Args:
+            results (dict): The resulting dict to be modified and passed
+                to the next transform in pipeline.
+        """
+        num_clips = results['num_clips']
+
+        if num_clips > 1 and (not self.is_multisampling):
+            self.is_multisampling = True
+            self.mean = np.repeat(self.mean, num_clips, axis=0)
```
Comment (innerlee) [inline mmaction/datasets/pipelines/augmentations.py:1312]: is this repeat really necessary?

File: configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+# dataset settings
+dataset_type = 'RawframeDataset'
+data_root = 'data/ucf101/rawframes'
+data_root_val = 'data/ucf101/rawframes'
+split = 1  # official train/test splits. valid numbers: 1, 2, 3
+ann_file_train = f'data/ucf101/ucf101_train_split_{split}_rawframes.txt'
+ann_file_val = f'data/ucf101/ucf101_val_split_{split}_rawframes.txt'
+ann_file_test = f'data/ucf101/ucf101_val_split_{split}_rawframes.txt'
+img_norm_cfg = dict(mean=[0, 0, 0], std=[1, 1, 1], to_bgr=False)
+# download c3d_mean_file from 'https://download.openmmlab.com/mmaction/'
```
Comment (innerlee) [inline configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:33]: curios why they do this. i saw the file has the shape `(3, 16, 128, 171)`, isn't this too much? will this tuned means applicable to other datasets?

so the resolution is also fixed by this? do you have resize if the input resolution are larger?

File: mmaction/models/backbones/c3d.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+    def __init__(self,
+                 pretrained=None,
+                 style='pytorch',
+                 conv_cfg=dict(type='Conv3d'),
+                 norm_cfg=None,
+                 act_cfg=dict(type='ReLU'),
+                 dropout_ratio=0.5,
+                 init_std=0.005,
+                 **kwargs):
```
Comment (innerlee) [inline mmaction/models/backbones/c3d.py:37]: is this `kwargs` used?

File: configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py (hunk 12 above)
Comment (congee524) [reply configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:33]: Thank you for your code review. I'm also confused about it. The file is not applicable to other datasets and it is the volume mean of UCF-101.

The relevant explanation is in the C3D doc "https://docs.google.com/document/d/1-QqZ3JHd76JfimY4QKqOojcEaf5g3JS0lNh-FHTxLag/"
image

However, I cannot find the definition of volume mean. From the compute file, the computing method seems traversing all `(16, 3, 128, 171)` clips in the dataset, and calculate the mean value of each position.

If I use the common image regularization method, the accuracy is 79.75 (40 epochs) and is smaller than 80.04 (30 epochs) using volume mean under the same parameter setting.

Comment (congee524) [reply configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:33]: So the resolution is also fixed. In the data pipeline, the image will be resized to (128, 171)

File: configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py (hunk 8 above)
Comment (congee524) [reply configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:7]: I used yapf, but I just switched to autopep8, and seems better.

Comment (innerlee) [reply configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:7]: okay it is yapf's fault: https://github.com/google/yapf/issues/423

let's keep it in two lines for now since many devs use yafp.

File: configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py (hunk 12 above)
Comment (congee524) [reply configs/recognition/c3d/c3d_sports1m_16x1x1_30e_ucf101_rgb.py:33]: I have removed the data augmentation method(subtract volume mean), since it is not so generic and the accuracy is not greatly improved.

**Comment by dreamerlin** [comment]
Also you can update the changelog as a `Highlights` and `New Features`

File: mmaction/models/__init__.py [hunk 14]
```diff
@@ -26,4 +27,14 @@
     'build_model', 'OHEMHingeLoss', 'SSNLoss', 'ResNet3dCSN', 'ResNetTIN',
     'TPN', 'TPNHead', 'build_loss', 'build_neck', 'AudioRecognizer',
     'AudioTSNHead'
+=======
```
Comment (dreamerlin) [inline mmaction/models/__init__.py:30]: conflict here

File: mmaction/models/backbones/c3d.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+        act_cfg (dict): Config dict for activation layer.
+            Default: ``dict(type='ReLU')``.
+        dropout_ratio (float): Probability of dropout layer. Default: 0.5.
+        init_std (float): Std value for Initiation of fc layers. Default: 0.01.
+    """
+
+    def __init__(self,
+                 pretrained=None,
+                 style='pytorch',
+                 conv_cfg=dict(type='Conv3d'),
```
Comment (innerlee) [inline mmaction/models/backbones/c3d.py:32]: do not use dict as default argument https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html#:~:text=Usually%20when%20a%20programmer%20uses,that%20the%20function%20is%20called.&text=The%20first%20time%20that%20the,for%20the%20list%20or%20dictionary.

Comment (dreamerlin) [reply mmaction/models/backbones/c3d.py:32]: Yeah, it is not good to use mutable variable as default argument. But there is no better way to correct it currently and it is used in many places in our Open-MMLab repo. Maybe we can keep it until we find a better way to replace it.

Comment (innerlee) [reply mmaction/models/backbones/c3d.py:32]: how does other repos handle this, like opencv, pytorch, etc?

Comment (congee524) [reply mmaction/models/backbones/c3d.py:32]: In pytorch, when need to use mutable variable as default arg. They set the arg as None.

and then
```
if arg is None:
    arg = [...]
```

Comment (innerlee) [reply mmaction/models/backbones/c3d.py:32]: two solutions:
1. use the pattern in the above link, i.e., use None as default, and assign it a dict in the function body
2. use current style, but protect the dict by deepcopy and adding a note: `self.act_cfg = cp.deepcopy(act_cfg)  # protect default argument`

File: mmaction/models/backbones/c3d.py [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+class C3D(nn.Module):
+    """C3D backbone.
+
+    Args:
+        pretrained (str | None): Name of pretrained model.
+        style (str): ``pytorch`` or ``caffe``. If set to "pytorch", the
+            stride-two layer is the 3x3 conv layer, otherwise the stride-two
+            layer is the first 1x1 conv layer. Default: 'pytorch'.
+        conv_cfg (dict): Config dict for convolution layer.
+            Default: ``dict(type='Conv3d')``.
```
Comment (dreamerlin) [inline mmaction/models/backbones/c3d.py:20]: Default: None

Comment (dreamerlin) [reply mmaction/models/backbones/c3d.py:20]:
```python
conv_cfg (dict | None): Config dict for convolution layer. If set to None, it will use ``dict(type='Conv3d')`` to construct layers. Default: None.
```

**Review by innerlee — APPROVED** [review]
